### ISSUE TITLE
4장

### ISSUE THREAD

**Issue by kangchanguk** [kangchanguk/study_ml#3]
# Chapter 4 model_train

## 선형 회귀

- 종속 변수 y와 한 개 이상의 독립 변수 (또는 설명 변수) X와의 선형 상관 관계를 모델링하는 회귀분석 기법

- y^ = h(x) = 파라미터 벡터 * 입력 특성(파라미터 벡터와 입력 특성의 점곱)

- 훈련: 훈련세트에 가장 맞는 모델 파라미터를 설정하는 것

### 정규 방정식

- 비용함수를 최소화 하는 특성 파라미터 값을 찾기 위한 해석적인 방법

- sklearn을 활용하여 선형 회귀

```
from sklearn.linear_model import LinearRegression

lin_reg = LinearRegression()
lin_reg.fit(x,y)

lin_reg.intercept_(절편), lin_reg.coef_(기울기), lin_reg.predict()
```

- 계산 복작도 적인 측면에서 역행렬을 직접 계산하는 것 보다 유사 역행렬을 구하는 방식이 더욱 빠르다

## 경사 하강법

- 비용함수를 최소화 하기위해 반복해서 파라미터를 조정해나가는 것
- 파라미터 벡터에 대해 비용 함수의 현재 그레이디언트(gradient)를 계산한 뒤 gradient가 감소하는 방향으로 진행
- 학습률:  스탭의 크기
     - 학습률이 너무 클 때: 알고리즘을 더 큰 값으로 발산하게 만듬
     - 학습률이 너무 작다면:  local minimum(지역 최솟값)에 수렴될 가능성이 높음

- 선형회귀의 MSE 비용함수는 곡선에서 어떤 두 점을 선택해 선을 그어도 곡선을 가로지르지 않는 볼록 함수
- 선형 회귀에 경사하강법을 적용하면 global minimum에 도달할 수 있음을 의미
- 특성 스케일을 적용한다면 global minimum에 도달하는 시간이 줄어든다.

### 배치 경사 하강법
- 경사 하강법을 구현하려면 각 모델 파라미터에 대해 비용 함수의 그레이디언트를 계산해야한다.
- 비용함수의 편미분을 사용하여 gradient에 따라 비용함수가 얼마나 바뀌는 지를 계산

![image](https://user-images.githubusercontent.com/26477881/125191484-a6abd300-e27d-11eb-9dc3-c74673b462ab.png)

- 훈련 세트 전체에 대해서 배치 하강 경사법을 실시, 속도가 느림 (full gradient descent)

### 학습률

- 적절한 학습률을 찾기 위해서 그리드 탐색을 활용
- 반복횟수를 크게 설정해두고 허용오차보다 작아지면 경사하강법이 최솟값에 도달한 것으로 간주

### 확률적 경사 하강법

- 배치 경사 하강법의 문제는 매 스탭 훈련세트 전체를 사용한다는 점
- 확률적 경사 하강법은 매 스탭에서 한 개의 샘플을 무작위로 선택하고 그 하나의 샘플에 대한 그레이디언트를 계산
- 매 스탭에서 다뤄야할 데이터가 적기 때문에 알고리즘이 빠름
- 확률적이어서 배치 경사 하강법보다 훨씬 불안정하다
- 무작위성을 보완하기 위해 learning scheduler를 도입(점진적으로 학습률을 감소시켜나감)
    - 학습률이 너무 빨리 줄어들면 local minimum에 갇히거나 global minimum까지 가는 중간에 stop
    - 학습률이 너무 천천히 줄어들면 오랫동안 최솟값 주변을 맴돌거나 훈련을 일찍 중단
- 샘플을 무작위로 선택하기 때문에 모든 샘플을 사용하게 하기위해 선택한 뒤 다시 샘플들을 섞음
```
n_epochs = 50
t0, t1 = 5, 50

def learning_scheduler(t):
    return t0 / (t+t1)

theta = np.random.randn(2,1)

for epoch in range(n_epochs):
    for i in range(m):
        random_index = np.random.randint(m)
        xi = x_b[random_index: random_index + 1]
        yi = y[random_index: random_index + 1]
        gradients = 2* xi.T.dot(xi.dot(theta) - yi)
        eta = learning_scheduler(epoch * + i)
        theta = theta - eta * gradients
```
- sklearn에서 SGD 방식으로 회귀하기 위해서는 SGDregressor 클래스를 활용
```
 from sklearn.linear_model import SGDRegressor

sgd_reg= SGDRegressor(max_iter = 1000, tol=1e-3, penalty = None, eta0=0.1)
sgd_reg.fit(x, y.ravel())
```

### 미니배치 경사 하강법
- 미니배치라 부르는 임의의 작은 샘플 세트에 대해 그레이디언트를 계산
- GPU(배치들에 대해 병렬 처리)

## 다항 회귀
- 각 특성의 거듭제곱을 새로운 특성으로 추가하여 비 선형 데이터를 학습하는데 선형모델을 사용
- 특성이 여러개일 때 다항 회귀는 이 특성 사이의 관계를 찾을 수 있음
- PolynomialFeatures가 주어진 차수까지 특성 간의 모든 교차항을 추가하기 때문
    - ex) a,b (degree = 3) = a^2, ab, a^2*b ....까지 특성으로 추가

```
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression

poly_features = PolynomialFeatures(degree=2, include_bias=False)
x_poly = poly_features.fit_transform(x)
lin_reg = LinearRegression()
lin_reg.fit(x_poly, y)from sklearn.preprocessing import PolynomialFeatures
```
